packages/engine/lib/stateful/src/field/source.rs

```diff
@@ -1,8 +1,45 @@
-use crate::error::Result;
+use core::fmt;
+use std::cmp::Ordering;
 
-pub trait FieldSource {
+use serde::Serialize;
+
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, Serialize)]
+pub struct PackageId(usize);
+
+impl PackageId {
+    #[inline]
+    pub fn as_usize(&self) -> usize {
+        self.0
+    }
+}
+
+impl fmt::Display for PackageId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+impl From<usize> for PackageId {
+    fn from(id: usize) -> Self {
+        Self(id)
+    }
+}
+
+/// Defines the source from which a Field was specified, useful for resolving clashes
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum FieldSource {
```

Comment:
Does anything else in the stateful crate know about Packages?

I don't really like this but I think it's a lot better than it was with everything being a generic.

Reply:
`Root` and `Component`?

Reply:
I think this is just stemming from some level of perfectionism and wanting us to have perfectly abstracted API boundaries. Maybe keeping it as Package is just clearer because we're not going to have another type of component anytime soon

Reply:
I feel the same, I'd love to have another name for this but I think this would be more confusing than helpful. 😒